### PR TITLE
webgpu: Optimize depthwise conv2d

### PR DESCRIPTION
**Pull request by qjia7**
This PR uses linear workgroup size to optimize depthwise conv2d vec4.

See 150% improvement for the time of FusedDepthwiseConv2D on
BlazePoseDetector.

To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

### CODE REVIEW

**Comment by qjia7** [comment]
See big improvement on Intel Gen9 GPU, for example DepthwiseConv2dNative in DeepLabV3 becomes 8.28ms from 21.06ms. No obvious changes for other GPUs.

File: tfjs-backend-webgpu/src/depthwise_conv2d_vec4_webgpu.ts [hunk 1]
```diff
@@ -82,11 +90,16 @@ export class DepthwiseConv2DVec4Program implements WebGPUProgram {
...
-        let batch = i32(globalId.z) / uniforms.outShape[1];
-        let r = i32(globalId.z) % uniforms.outShape[1];
-        let c = i32(globalId.y) * ${this.workPerThread};
-        let d1 = i32(globalId.x) * 4;
+      ${main('index')} {
+        let width0 = uniforms.outShape[3] / 4;
+        let d1 = (index % width0) * 4;
+        var index1 = index / width0;
+        let width1 = uniforms.virtualWidth / ${this.workPerThread};
+        let c = (index1 %  width1) * ${this.workPerThread};
```
Comment (gyagp) [inline tfjs-backend-webgpu/src/depthwise_conv2d_vec4_webgpu.ts:98]:
```suggestion
        let c = (index1 % width1) * ${this.workPerThread};
```

**Review by gyagp — APPROVED** [review]
LGTM % nit

**Comment by qjia7** [comment]
depthwise_opt
I draw a picture to show the differences. In above picture, the shadow places are the data that threads in the same workgroup will process. We can see that linear workgroup size makes the processed data more contiguous in memory for threads in one workgroup. I guess it may be the reason that we see big improvement on Gen9 GPU. There are no obvious changes for other GPUs. Maybe other drivers can well handle this situation.

**Review by xhcao — APPROVED** [review]
LGTM. But I feel it is weird to introduce the new `virtualWidth` concept and the complicated computation of coordinate here. I guess the optimization is valid when `channels` is large or low-end devices.

**Comment by qjia7** [comment]
> LGTM. But I feel it is weird to introduce the new `virtualWidth` concept and the complicated computation of coordinate here. I guess the optimization is valid when `channels` is large or low-end devices.

Thank you Xinghua. The `virtualWidth` is to make sure the correctness of the algorithm when switching to linear workgroup size. In current algorithm, each thread writes 4x4 data which requires the 4 data in the same col, the 4x4 data in the same row. It means that the width of the output shape needs to be divisible by 4. That's why I introduced the `virtualWidth = Math.ceil(this.outputShape[2] / this.workPerThread) * this.workPerThread; ` where `this.workPerThread = 4`.

**Pull request merged**